File: mg01/game/tech/Assets/Script/Org/Action/Interval/ActionInterval.cs
/* ==============================================================================
 * 时序动作
 * @author jr.zeng
 * 2016/10/26 11:41:03
 * ==============================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using UnityEngine;

namespace mg.org.Actions
{

    public class ActionInterval : ActionBase
    {

        protected float m_duration = 1;
        protected float m_perDelta = 1;

        protected float m_elapsed = 0;

        protected float m_progress = 0;	//当前进度

        public ActionInterval()
        {

        }

        public void InitWithDuration(float duration_)
        {
            Duration = duration_;

            m_inited = true;
        }

        protected override void OnReset()
        {
            m_elapsed = 0;
            m_progress = 0;
        }

        public override void Step(float dt_)
        {

            m_elapsed += dt_;

            float progress = 1;
            if (m_duration > 0)
            {
                progress = m_elapsed * m_perDelta;
            }
            
            Progress(progress);

        }


        public override void Progress(float progress_)
        {
            m_progress = Mathf.Clamp01(progress_);

            OnProgress(m_progress);

            if (m_progress >= 1)
            {
                Done();
            }
        }


        virtual protected void OnProgress(float progress_)
        {

        }

        override public float Duration
        {
            get  { return m_duration; }
            set
            {
                m_duration = value;
                if (m_duration > 0)
                {
                    m_perDelta = Mathf.Abs(1f / m_duration) * Mathf.Sign(m_perDelta);
                }
                else
                {
                    m_duration = 0;
                    m_perDelta = float.MaxValue;
                }

            }
        }

        public float Elapsed        { get { return m_elapsed; } }

      
        protected override void OnClear()
        {
            m_elapsed = 0;
            m_progress = 0;

        }


    }

}

File: mg01/game/tech/Assets/Resources/LuaScript/src/org/action/interval/abs/ActionInterval.lua
--ActionInterval(抽象)
--@author jr.zeng
--2017年7月14日 下午2:28:14
local modname = "ActionInterval"
--==================global reference=======================

--===================namespace========================
local ns = "org.action"
local using = {--[[namespace1,...]]}
local _ENV = namespace(ns, using)
--===================module property========================
local super = ActionBase	--父类
_ENV[modname] = class(modname, super)
local ActionInterval = _ENV[modname]

local clamp = MathUtil.clamp

--===================module content========================

function ActionInterval:__ctor()
    
    self.m_duration = 1 --周期
    self.m_perDelta = 1 --周期系数
    
    self.m_elapsed = 0  --经过时间
    self.m_progress = 0 --当前进度
    
end


function ActionInterval:initWithDuration(duration)
    
    self:setDuration(duration)
    self.m_inited = true
end


--设置动画时间
function ActionInterval:setDuration(duration)

	self.m_duration = duration
	if duration > 0 then
		self.m_perDelta = 1 / self.m_duration
	else
		self.m_duration = 0
		self.m_perDelta = 0
	end
end

--override
function ActionInterval:getDuration()
	return self.m_duration
end


--获取经过时间
function ActionInterval:getElapsed()
	return self.m_elapsed
end



--//-------~★~-------~★~-------~★~数据管理~★~-------~★~-------~★~-------//


function ActionInterval:update(dt)
    
    self.m_elapsed = self.m_elapsed + dt
    
    local progress = 1
    if self.m_duration > 0 then
        progress = self.m_elapsed * self.m_perDelta
    end
    
    self:progress(progress)
end

--override
function ActionInterval:progress(value)

    --更新进度
    value = clamp(value)
    self.m_progress = value
    
    self:__progress(value)
    
    if value >= 1 then
        self:done()
    end
end

--virtual
function ActionInterval:__progress(value)

end



--//-------~★~-------~★~-------~★~数据操作~★~-------~★~-------~★~-------//


--override
function ActionInterval:__reset()

    self.m_elapsed = 0
    self.m_progress = 0
end

--override
function ActionInterval:__clear()

    self.m_elapsed = 0
    self.m_progress = 0
end

return ActionInterval